File: cms-frontend/src/app/services/enrollment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Enrollment } from '../models/enrollment.model';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class EnrollmentService {
  private readonly apiUrl = `${environment.apiUrl}/enrollments`;

  constructor(private readonly http: HttpClient) { }

  getAllEnrollments(): Observable<Enrollment[]> {
    return this.http.get<Enrollment[]>(this.apiUrl);
  }

  getEnrollmentById(id: number): Observable<Enrollment> {
    return this.http.get<Enrollment>(`${this.apiUrl}/${id}`);
  }

  createEnrollment(studentId: string, courseCode: string, semester: string, academicYear: string): Observable<Enrollment> {
    const payload = { studentId, courseCode, semester, academicYear };
    return this.http.post<Enrollment>(this.apiUrl, payload);
  }

  updateEnrollmentStatus(id: number, status: string): Observable<Enrollment> {
    return this.http.put<Enrollment>(`${this.apiUrl}/${id}/status`, { status });
  }

  updateEnrollmentGrade(id: number, grade: string): Observable<Enrollment> {
    return this.http.put<Enrollment>(`${this.apiUrl}/${id}/grade`, { grade });
  }

  deleteEnrollment(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getEnrollmentsByStudent(studentId: string): Observable<Enrollment[]> {
    return this.http.get<Enrollment[]>(`${this.apiUrl}/student/${studentId}`);
  }

  getEnrollmentsByCourse(courseCode: string): Observable<Enrollment[]> {
    return this.http.get<Enrollment[]>(`${this.apiUrl}/course/${courseCode}`);
  }
}

File: cms-backend/src/main/java/com/thishan/cms_backend/service/EnrollmentService.java
package com.thishan.cms_backend.service;

import com.thishan.cms_backend.entity.Enrollment;
import com.thishan.cms_backend.entity.Course;
import com.thishan.cms_backend.entity.Student;
import com.thishan.cms_backend.repository.EnrollmentRepository;
import com.thishan.cms_backend.repository.CourseRepository;
import com.thishan.cms_backend.repository.StudentRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class EnrollmentService {
    
    @Autowired
    private EnrollmentRepository enrollmentRepository;
    
    @Autowired
    private CourseRepository courseRepository;
    
    @Autowired
    private StudentRepository studentRepository;
    
    // Get all enrollments
    public List<Enrollment> getAllEnrollments() {
        return enrollmentRepository.findAll();
    }
    
    // Get enrollment by ID
    public Optional<Enrollment> getEnrollmentById(Long id) {
        return enrollmentRepository.findById(id);
    }
    
    // Get enrollments by student ID
    public List<Enrollment> getEnrollmentsByStudentId(String studentId) {
        return enrollmentRepository.findByStudentId(studentId);
    }
    
    // Get enrollments by course code
    public List<Enrollment> getEnrollmentsByCourseCode(String courseCode) {
        return enrollmentRepository.findByCourseCode(courseCode);
    }
    
    // Create a new enrollment
    public Enrollment createEnrollment(String studentId, String courseCode, String semester, String academicYear) {
        // Find student
        Student student = studentRepository.findByStudentId(studentId)
                .orElseThrow(() -> new RuntimeException("Student not found with ID: " + studentId));
        
        // Find course
        Course course = courseRepository.findByCode(courseCode)
                .orElseThrow(() -> new RuntimeException("Course not found with code: " + courseCode));
        
        // Check if student is already enrolled in this course
        if (enrollmentRepository.existsByStudentIdAndCourseCode(studentId, courseCode)) {
            throw new RuntimeException("Student " + studentId + " is already enrolled in course " + courseCode);
        }
        
        // Create new enrollment
        Enrollment enrollment = new Enrollment(student, course, semester, academicYear);
        return enrollmentRepository.save(enrollment);
    }
    
    // Update enrollment status
    public Enrollment updateEnrollmentStatus(Long enrollmentId, Enrollment.EnrollmentStatus status) {
        Enrollment enrollment = enrollmentRepository.findById(enrollmentId)
                .orElseThrow(() -> new RuntimeException("Enrollment not found with id: " + enrollmentId));
        
        enrollment.setStatus(status);
        return enrollmentRepository.save(enrollment);
    }
    
    // Update enrollment grade
    public Enrollment updateEnrollmentGrade(Long enrollmentId, String grade) {
        Enrollment enrollment = enrollmentRepository.findById(enrollmentId)
                .orElseThrow(() -> new RuntimeException("Enrollment not found with id: " + enrollmentId));
        
        enrollment.setGrade(grade);
        // If grade is assigned, mark as completed
        if (grade != null && !grade.trim().isEmpty()) {
            enrollment.setStatus(Enrollment.EnrollmentStatus.COMPLETED);
        }
        
        return enrollmentRepository.save(enrollment);
    }
    
    // Delete an enrollment
    public void deleteEnrollment(Long id) {
        Enrollment enrollment = enrollmentRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Enrollment not found with id: " + id));
        enrollmentRepository.delete(enrollment);
    }
    
    // Get enrollments by status
    public List<Enrollment> getEnrollmentsByStatus(Enrollment.EnrollmentStatus status) {
        return enrollmentRepository.findByStatus(status);
    }
    
    // Get enrollments by semester
    public List<Enrollment> getEnrollmentsBySemester(String semester) {
        return enrollmentRepository.findBySemester(semester);
    }
    
    // Get enrollments by academic year
    public List<Enrollment> getEnrollmentsByAcademicYear(String academicYear) {
        return enrollmentRepository.findByAcademicYear(academicYear);
    }
    
    // Get students enrolled in a course
    public List<Student> getStudentsInCourse(String courseCode) {
        return enrollmentRepository.findStudentsByCourseCode(courseCode);
    }
    
    // Get courses enrolled by a student
    public List<Course> getCoursesForStudent(String studentId) {
        return enrollmentRepository.findCoursesByStudentId(studentId);
    }
    
    // Get enrollment count for a course
    public Long getEnrollmentCountForCourse(String courseCode) {
        return enrollmentRepository.countByCourseCode(courseCode);
    }
    
    // Check if student is enrolled in course
    public boolean isStudentEnrolledInCourse(String studentId, String courseCode) {
        return enrollmentRepository.existsByStudentIdAndCourseCode(studentId, courseCode);
    }
}
